Issue #929: Use private final loggers instead of private static final in Checkstyle codebase

crud3 opened the pull request:
Issue #929

There is also an (abandoned and closed) PR for this issue: #4076

I have tried to take both the discussion from the issue and the incomplete PR into consideration.

Changes:
- Using private final loggers instead of private static final, except in the Main.java class
- Added sventu check to prevent static Loggers in the future
- Added exception to above rule for the Main.java class (for reasons see discussion in #4076)

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/Checker.java:
```diff
@@ -68,7 +68,8 @@
     /** Logger for Checker. */
-    private static final Log LOG = LogFactory.getLog(Checker.class);
+    @SuppressWarnings("PMD.LoggerIsNotStaticFinal")
+    private final Log log;
```
@crud3 , please follow violations from CIs, all of them should be satisfied (green).

it is not allowed to use PMD suppressions by annotation.
We need to disable such PMD rule completely in our project as we are library and static is preferable in final applications only.

crud3 replied:
Thanks for the feedback, I will get to work on it tomorrow!

crud3 replied:
@romani I have removed the the PMD suppression annotations, squashed and rebased.
I also got rid of another Error by adding MainTest.java to the sevntu_suppressions.xml like Main.java
`    <!-- Main and MainTest are technically not part of the checkstyle library,
    so static Loggers are okay there. -->
`
`    <suppress checks="AvoidModifiersForTypesCheck"
              files="Main\.java|MainTest\.java"/>`
to suppress this rule in checkstyle_sevntu_checks.xml:
`<module name="AvoidModifiersForTypesCheck">`
`          <property name="forbiddenClassesRegexpProtected" value="Logger"/>`
`          <property name="forbiddenClassesRegexpPublic" value="Logger"/>`
`          <property name="forbiddenClassesRegexpPackagePrivate" value="Logger"/>`
`          <property name="forbiddenClassesRegexpStatic" value="Logger"/>`
`</module>`

Is that okay?

Next I could add the changes in checkstyle_sevntu_checks.xml you proposed, to see if they get rid of the remaining 8 errors.

romani commented:
@rnveach ,
please pay attention to this PR, looks like we need to update configuration for 

in build log:
```
[INFO] There are 9 errors reported by Checkstyle 8.2 with /home/travis/build/checkstyle/checkstyle/config/checkstyle_sevntu_checks.xml ruleset.

[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java:[136,5] (extension) CustomDeclarationOrder: Field definition in wrong order. Expected 'Field(static final)' then 'Field(private final)'.
```

to my mind/experience it is better to place log/logger to the top of class to not mix it logic, as it is system related not a logic related. 

I propose following change to [checkstyle_sevntu_checks.xml](https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle_sevntu_checks.xml#L153)
```
$ git diff
diff --git a/config/checkstyle_sevntu_checks.xml b/config/checkstyle_sevntu_checks.xml
index 0246c22..c0b1cb7 100644
--- a/config/checkstyle_sevntu_checks.xml
+++ b/config/checkstyle_sevntu_checks.xml
@@ -152,7 +152,8 @@
         <module name="ForbidWildcardAsReturnType"/>
         <module name="CustomDeclarationOrder">
             <property name="customDeclarationOrder"
-                      value="InnerEnum(public .*)
+                      value="Field((private) final Log ([\w]*L|l)og)
+                      ### InnerEnum(public .*)
                       ### Field(public static final)

```

crud3 commented:
@romani I have removed the the PMD suppression annotations, squashed and rebased.
I also got rid of another Error by adding MainTest.java to the sevntu_suppressions.xml like Main.java
`    <!-- Main and MainTest are technically not part of the checkstyle library,
    so static Loggers are okay there. -->
`
`    <suppress checks="AvoidModifiersForTypesCheck"
              files="Main\.java|MainTest\.java"/>`
to suppress this rule in checkstyle_sevntu_checks.xml:
`<module name="AvoidModifiersForTypesCheck">`
`          <property name="forbiddenClassesRegexpProtected" value="Logger"/>`
`          <property name="forbiddenClassesRegexpPublic" value="Logger"/>`
`          <property name="forbiddenClassesRegexpPackagePrivate" value="Logger"/>`
`          <property name="forbiddenClassesRegexpStatic" value="Logger"/>`
`</module>`

Is that okay?

Next I could add the changes in checkstyle_sevntu_checks.xml you proposed, to see if they get rid of the remaining 8 errors.

romani commented:
@crud3 , you are on right path.
I still see 2 commits, please squash them in one and proceed with CI fixes.

crud3 commented:
@romani I have squashed my commits and rebased. It passes all checks now and I think the original Issue is resolved with this.

However, I couldn't get the addition you proposed to work (change of the CustomDeclarationOrder to move the private final Logger at the top of the class), so I have left them at their respective places for the moment.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/Checker.java:
```diff
@@ -65,11 +65,12 @@
  * @author Andrei Selkin
  */
 public class Checker extends AutomaticBean implements MessageDispatcher, RootModule {
+
```
please remove this change, lets keep only required changes.

crud3 commented:
I have removed the newline.

Regarding the position of the logger field: I do not have a fix already. I tried to incorporate the changes you proposed to checkstyle_sevntu_checks.xml (and also tried some variations of it) and moved the logger to the top of the class but it produced checkstyle errors saying that the static fields that came after the logger declaration in the classes Checker.java and TranslationCheck.java are in the wrong order. I could push again on this PR so you can see them. I can also just create a new issue for it like you said and if no one wants to fix it in the course of Hacktoberfest I'll get to work on it in November.

Thanks for your guidance!

romani commented:
New issue is created https://github.com/checkstyle/checkstyle/issues/5209

@crud3 , thanks a lot for a fix, and you are always welcome with any other fix, just check for comments like "I am on it" at issue to see is it occupied or not
